mithril-aggregator/src/services/usage_reporter.rs

```diff
@@ -0,0 +1,307 @@
+        let message = &received_messages[0];
+        assert_eq!(message.source, "Metrics");
+        assert_eq!(message.action, metric.name());
+        let message_content: MetricMessage =
+            serde_json::from_value(message.content.clone()).unwrap();
+        assert_eq!(3, message_content.counter);
+        assert_eq!(Duration::from_secs(10), message_content.duration);
+    }
+
+    #[tokio::test]
+    async fn send_one_message_for_each_non_zero_value_metrics() {
+        let (mut usage_reporter, metric_service, mut rx) = build_usage_reporter();
+
+        let metric_1 = metric_service.get_certificate_total_produced_since_startup();
+        let metric_2 = metric_service.get_certificate_detail_total_served_since_startup();
+        metric_1.increment();
+        metric_2.increment();
+
+        usage_reporter.send_metrics(&Duration::from_secs(10));
+
+        let received_messages = received_messages(&mut rx);
+        let metric_values: Vec<(String, i64)> =
+            received_messages.iter().map(extract_metric_value).collect();
+
+        assert!(metric_values.contains(&(metric_1.name(), 1)));
+        assert!(metric_values.contains(&(metric_2.name(), 1)));
+    }
+
+    #[tokio::test]
+    async fn resend_only_delta_since_last_send() {
+        let (mut usage_reporter, metric_service, mut rx) = build_usage_reporter();
+
+        let metric_1 = metric_service.get_certificate_total_produced_since_startup();
+        let metric_2 = metric_service.get_certificate_detail_total_served_since_startup();
+
+        {
+            metric_1.increment_by(12);
+            metric_2.increment_by(5);
+            usage_reporter.send_metrics(&Duration::from_secs(10));
+
+            let received_messages = received_messages(&mut rx);
+            assert_eq!(2, received_messages.len());
+        }
+        {
+            metric_2.increment_by(20);
+            metric_2.increment_by(33);
+            usage_reporter.send_metrics(&Duration::from_secs(10));
+
+            let received_messages = received_messages(&mut rx);
+            assert_eq!(1, received_messages.len());
+
+            assert_eq!(
+                (metric_2.name(), 53),
+                extract_metric_value(&received_messages[0])
+            );
+        }
+        {
+            metric_2.increment_by(15);
+            usage_reporter.send_metrics(&Duration::from_secs(10));
+
+            let received_messages = received_messages(&mut rx);
+            assert_eq!(1, received_messages.len());
+
+            assert_eq!(
+                (metric_2.name(), 15),
+                extract_metric_value(&received_messages[0])
+            );
+        }
+    }
+
+    mod metric_delta {
+        fn build_hashmap<T: Copy>(values: &[(&str, T)]) -> HashMap<String, T> {
+            let mut metrics = HashMap::new();
+            for (name, value) in values {
+                metrics.insert(name.to_string(), *value);
+            }
+            metrics
+        }
```

Comment:
I can use a `into_iter` but I can't do better because of the `&str` to transform into a `String`.
I propose the following code. Do you have an other suggestion ?
```
        fn build_hashmap<T: Copy>(values: &[(&str, T)]) -> HashMap<String, T> {
            values
                .into_iter()
                .map(|(name, value)| (name.to_string(), *value))
                .collect()
        }
```
Note that is just a utility function to simplify writing tests in this file.

Reply:
Ok that works :+1:

Reply:
Done